File: fe2-android/app/src/main/java/com/github/dedis/popstellar/model/objects/security/KeyPair.java
package com.github.dedis.popstellar.model.objects.security;

import androidx.annotation.NonNull;

import com.github.dedis.popstellar.model.Immutable;
import com.google.crypto.tink.PublicKeySign;

import java.security.GeneralSecurityException;
import java.util.Objects;

/**
 * Represent a private/public key pair used for signing data
 *
 * <p>This object does not actually store the private key as is it safely stored by another module.
 */
@Immutable
public class KeyPair {

  private final PrivateKey privateKey;
  private final PublicKey publicKey;

  public KeyPair(PrivateKey privateKey, PublicKey publicKey) {
    this.privateKey = privateKey;
    this.publicKey = publicKey;
  }

  /**
   * Signs the given data with the private key of the pair {@link PublicKeySign#sign(byte[])}
   *
   * @param data to sign
   * @return the signature
   * @throws GeneralSecurityException if an error occurs
   */
  public Signature sign(Base64URLData data) throws GeneralSecurityException {
    return privateKey.sign(data);
  }

  public boolean verify(Signature signature, Base64URLData data) {
    return publicKey.verify(signature, data);
  }

  public PrivateKey getPrivateKey() {
    return privateKey;
  }

  public PublicKey getPublicKey() {
    return publicKey;
  }

  @NonNull
  @Override
  public String toString() {
    return getClass().getSimpleName()
        + "{privateKey="
        + privateKey.toString()
        + ", publicKey="
        + publicKey.toString()
        + "}";
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }

    KeyPair keyPair = (KeyPair) o;
    return Objects.equals(privateKey, keyPair.privateKey)
        && Objects.equals(publicKey, keyPair.publicKey);
  }

  @Override
  public int hashCode() {
    return Objects.hash(privateKey, publicKey);
  }
}

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/model/objects/security/KeyPair.kt
package com.github.dedis.popstellar.model.objects.security

import com.github.dedis.popstellar.model.Immutable
import java.security.GeneralSecurityException
import java.util.Objects

/**
 * Represent a private/public key pair used for signing data
 *
 * This object does not actually store the private key as is it safely stored by another module.
 */
@Immutable
open class KeyPair(open val privateKey: PrivateKey, val publicKey: PublicKey) {

  /**
   * Signs the given data with the private key of the pair [PublicKeySign.sign]
   *
   * @param data to sign
   * @return the signature
   * @throws GeneralSecurityException if an error occurs
   */
  @Throws(GeneralSecurityException::class)
  fun sign(data: Base64URLData): Signature {
    return privateKey.sign(data)
  }

  fun verify(signature: Signature, data: Base64URLData): Boolean {
    return publicKey.verify(signature, data)
  }

  override fun toString(): String {
    return "${javaClass.simpleName}{privateKey=$privateKey, publicKey=$publicKey}"
  }

  override fun equals(other: Any?): Boolean {
    if (this === other) {
      return true
    }
    if (other == null || javaClass != other.javaClass) {
      return false
    }
    val keyPair = other as KeyPair
    return privateKey == keyPair.privateKey && publicKey == keyPair.publicKey
  }

  override fun hashCode(): Int {
    return Objects.hash(privateKey, publicKey)
  }
}
